src/runtime_src/core/common/xrt_xclbin.cpp

```diff
@@ -62,15 +67,342 @@ int
 xrtXclbinUUID(xclDeviceHandle dhdl, xuid_t out)
 {
   try {
-   api::xrtXclbinUUID(dhdl, out);
-   return 0;
+    api::xrtXclbinUUID(dhdl, out);
+    return 0;
   }
-  catch (const xrt_core::error& ex) {
+  catch (const xrt_core::error &ex) {
     xrt_core::send_exception_message(ex.what());
     return ex.get();
   }
-  catch (const std::exception& ex) {
+  catch (const std::exception &ex) {
+    xrt_core::send_exception_message(ex.what());
+    return -1;
+  }
+}
+
+namespace xrt {
+
+// class xclbin_impl - class for xclbin objects
+//
+// Life time of xclbin are managed through shared pointers.
+// A buffer is freed when last references is released.
+class xclbin_impl
+{
+protected:
+  std::vector<char> m_axlf;
+  const axlf *m_top;
+
+  void init_axlf_handle()
+  {
+    const axlf *tmp = reinterpret_cast<const axlf *>(this->m_axlf.data());
+    if (strncmp(tmp->m_magic, "xclbin2", 7)) // Future: Do not hardcode "xclbin2"
+        throw std::runtime_error("Invalid xclbin");
+    this->m_top = tmp;
+  }
+
+public:
+  xclbin_impl(const std::vector<char> &data)
+  {
+    this->m_axlf = data;
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  xclbin_impl(const std::string &filename)
+  {
+    if (filename.empty())
+      throw std::runtime_error("No XCLBIN specified");
+
+    // Load the file into this->data
+    std::ifstream stream(filename);
+    stream.seekg(0, stream.end);
+    size_t size = stream.tellg();
+    stream.seekg(0, stream.beg);
+    this->m_axlf.resize(size);
+    stream.read(this->m_axlf.data(), size);
+
+    // Set pointer of type axlf*, and check magic string
+    init_axlf_handle();
+  }
+
+  void
+  check_empty() const
+  {
+    if (this->m_axlf.size() == 0 || this->m_top == nullptr)
+      throw xrt_core::error(-EINVAL, "Invalid XCLBIN data");
+  }
+
+  std::vector<std::string>
+  get_cu_names() const
+  {
+    std::vector<std::string> names;
+    for (auto cu : xrt_core::xclbin::get_cus(m_top, false)) // Why would I ever need them encoded
+      names.push_back(xrt_core::xclbin::get_ip_name(m_top, cu));
+
+    return names;
+  }
+
+  std::string
+  get_xsa_name() const
+  {
+    check_empty();
+    return reinterpret_cast<const char *>(m_top->m_header.m_platformVBNV);
+  }
+
+  uuid
+  get_uuid() const
+  {
+    check_empty();
+    return uuid(m_top->m_header.uuid);
```

Comment:
```suggestion
    return { m_top->m_header.uuid };
```
or perhaps even without `{}`

Reply:
Will fix in next push